service/controller/machinepoolexp/handler/migration/resource_test.go

```diff
@@ -0,0 +1,234 @@
+		return nil, microerror.Mask(err)
+	}
+
+	switch t.Kind {
+	case "MachinePool":
+		obj = new(oldcapiexpv1alpha3.MachinePool)
+	case "AzureMachinePool":
+		obj = new(oldcapzexpv1alpha3.AzureMachinePool)
+	default:
+		return nil, microerror.Maskf(unknownKindError, "kind: %s", t.Kind)
+	}
+
+	// ...and unmarshal the whole object.
+	err = yaml.Unmarshal(bs, obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return obj, nil
+}
+
+func newFakeClient() client.Client {
+	scheme := runtime.NewScheme()
+
+	// Upstream CAPI schema
+	err := capiexp.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+
+	// Upstream CAPZ schema
+	err = capzexp.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+
+	// Old experimental CAPI schema
+	err = oldcapiexpv1alpha3.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+
+	// Old experimental CAPZ schema
+	err = oldcapzexpv1alpha3.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+
+	return fake.NewClientBuilder().WithScheme(scheme).Build()
+}
+
+// normalizeFileName converts all non-digit, non-letter runes in input string to
+// dash ('-'). Coalesces multiple dashes into one.
+func normalizeFileName(s string) string {
+	var result []rune
+	for _, r := range s {
+		if unicode.IsDigit(r) || unicode.IsLetter(r) {
+			result = append(result, r)
+		} else {
+			l := len(result)
+			if l > 0 && result[l-1] != '-' {
+				result = append(result, rune('-'))
+			}
+		}
+	}
+	return string(result)
+}
+
+func verifyCR(t *testing.T, client client.Client, testName string, obj client.Object, nsName types.NamespacedName) {
+	t.Helper()
+
+	err := client.Get(context.Background(), nsName, obj)
+	if err != nil {
+		t.Fatalf("err = %#q, want %#v", microerror.JSON(err), nil)
+	}
+
+	ta, err := meta.TypeAccessor(obj)
+	if err != nil {
+		t.Fatalf("err = %#q, want %#v", microerror.JSON(err), nil)
+	}
+
+	out, err := yaml.Marshal(obj)
+	if err != nil {
+		t.Fatalf("err = %#q, want %#v", microerror.JSON(err), nil)
+	}
+
+	p := filepath.Join("testdata", fmt.Sprintf("%s_%s.golden", normalizeFileName(testName), ta.GetKind()))
+
+	if *update {
+		err = ioutil.WriteFile(p, out, 0644) // nolint:gosec
+		if err != nil {
+			t.Fatalf("err = %#q, want %#v", microerror.JSON(err), nil)
+		}
+	}
+
+	goldenFile, err := ioutil.ReadFile(p)
+	if err != nil {
+		t.Fatalf("err = %#q, want %#v", microerror.JSON(err), nil)
+	}
+
+	//
+	// XXX: Workaround to exclude timestamp differences from comparison.
+	//
+	re := regexp.MustCompile(`lastTransitionTime: "\d+-\d+-\d+T\d+:\d+:\d+Z"`)
+	timeNow := time.Now()
+	goldenFile = re.ReplaceAll(goldenFile, []byte(fmt.Sprintf(`lastTransitionTime: "%s"`, timeNow)))
+	out = re.ReplaceAll(out, []byte(fmt.Sprintf(`lastTransitionTime: "%s"`, timeNow)))
+
+	// Final comparison of golden version vs. one generated by test.
+	if !bytes.Equal(out, goldenFile) {
+		t.Fatalf("\n\n%s\n", cmp.Diff(string(goldenFile), string(out)))
+	}
+}
+
+func TestEnsureCreatedIsIdempotent(t *testing.T) {
```

Comment:
Test that checks if the EnsureCreated is idempotent, it calls the function two times. The EnsureCreated func is creating CRs, so create call would fail if done more than once.